### PR TITLE
Making create_snapshot output more informative by adding more fields

### PR DESCRIPTION
**Pull request by venu-block1**
## Change Description
Making create_snapshot output more informative by adding more fields -- head_block_num, head_block_time, and snapshot_version.

## Change Type
**Select ONE**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

The API Reference documentation for create_snapshot -- https://developers.eos.io/manuals/eos/latest/nodeos/plugins/producer_api_plugin/api-reference/index#operation/create_snapshot should be updated for 3 new fields:
        head_block_num:
                      type: integer
                      description: Highest block number on the chain
        head_block_time:
                      type: string
                      description: Highest block unix timestamp
        version:
                      type: integer
                      description: Version number

### CODE REVIEW

File: plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp [hunk 1]
```diff
@@ -44,6 +44,9 @@ class producer_plugin : public appbase::plugin<producer_plugin> {
 
    struct snapshot_information {
       chain::block_id_type head_block_id;
+      uint32_t             head_block_num;
```
Comment (linhuang-blockone) [inline plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:47]: So the new fields are only added to snapshot_information of producer_plugin? At first I thought the snapshot format itself would be changed (which should be avoided).

**Review by linhuang-blockone — COMMENTED** [review]
Can you add some unit test cases for this?

File: plugins/producer_plugin/producer_plugin.cpp [hunk 2]
```diff
@@ -150,7 +162,7 @@ class pending_snapshot {
                  ("ec", ec.value())
                  ("message", ec.message()));
 
-      return {block_id, final_path};
+      return {block_id, chain.head_block_num(), chain.head_block_time(), snapshot_version(final_path), final_path};
```
[the remark this answers is not in the document]
Comment (venu-block1) [reply plugins/producer_plugin/producer_plugin.cpp]: Made changes accordingly. Please review.

[the remark this answers is not in the document]
Comment (venu-block1) [reply plugins/producer_plugin/producer_plugin.cpp]: Done

File: plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp (hunk 1 above)
Comment (venu-block1) [reply plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:47]: Correct, changes are confined to producer_plugin.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -3184,6 +3184,16 @@ std::optional<uint64_t> controller::convert_exception_to_error_code( const fc::e
    return e_ptr->error_code;
 }
 
+chain_snapshot_header controller::extract_chain_snapshot_header(snapshot_reader& snapshot) {
```
[the remark this answers is not in the document]
Comment (venu-block1) [reply libraries/chain/controller.cpp]: The parameter should have been made const, but the function is a static function, and cannot be const. In any case, this function is not needed any more once I implemented changes for other review comments.

File: unittests/snapshot_tests.cpp [hunk 4]
```diff
@@ -808,4 +812,56 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(test_kv_snapshot, SNAPSHOT_SUITE, snapshot_suites)
    }
 }
 
+using next_t = eosio::producer_plugin::next_function<eosio::producer_plugin::snapshot_information>;
+
+void next_func(eosio::producer_plugin::snapshot_information& snap_info)
```
Comment (linhuang-blockone) [inline unittests/snapshot_tests.cpp:817]: I don't see next_func is used in this file.

File: unittests/snapshot_tests.cpp [hunk 5]
```diff
@@ -808,4 +812,56 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(test_kv_snapshot, SNAPSHOT_SUITE, snapshot_suites)
...
+   };
+
+   auto final_path = eosio::pending_snapshot::get_final_path(block2->previous, "../snapshots/");
+   auto pending_path = eosio::pending_snapshot::get_pending_path(block2->previous, "../snapshots/");
+
+   write_snapshot( pending_path );
+   next_t next;
+   eosio::pending_snapshot pending{ block2->previous, next, pending_path.generic_string(), final_path.generic_string() };
+   test_snap_info = pending.finalize(*chain.control);
+   BOOST_REQUIRE_EQUAL(test_snap_info.head_block_num, 6);
```
Comment (linhuang-blockone) [inline unittests/snapshot_tests.cpp:863]: Is this always 6?

File: unittests/snapshot_tests.cpp (hunk 4 above)
Comment (venu-block1) [reply unittests/snapshot_tests.cpp]: Good catch. I cleaned up the unused code.

File: unittests/snapshot_tests.cpp (hunk 5 above)
Comment (venu-block1) [reply unittests/snapshot_tests.cpp:858]: Yes, it is always 6.

File: unittests/CMakeLists.txt [hunk 6]
```diff
@@ -42,7 +42,7 @@ add_custom_command(
 file(GLOB UNIT_TESTS "*.cpp") # find all unit test suites
 add_executable( unit_test ${UNIT_TESTS} protocol_feature_digest_tests.cpp) # build unit tests as one executable
 
-target_link_libraries( unit_test eosio_chain_wrap chainbase eosio_testing fc appbase state_history abieos ${PLATFORM_SPECIFIC_LIBS} )
+target_link_libraries( unit_test eosio_chain_wrap chainbase eosio_testing fc appbase state_history producer_plugin abieos ${PLATFORM_SPECIFIC_LIBS} )
```
Comment (larryk85) [inline unittests/CMakeLists.txt:45]: For the unit tests, we don't really want them in the set of core unit tests for eos.  If you take a look at `trace_api_plugin` or `resource_monitor_plugin` for reference on what the file structure and CMake, etc. look like.

But, basically:
In producer_plugin directory, create a folder called `test` then add_subdirectory that in the main CMakeLists there.
You can then copy the structure of one of the test cpp files there and place your unit tests below there.

**Review by larryk85 — CHANGES_REQUESTED** [review]

File: unittests/CMakeLists.txt (hunk 6 above)
Comment (venu-block1) [reply unittests/CMakeLists.txt]: Moved the unit test from core unit tests to a new `test` directory in `producer_plugin`.

**Review by larryk85 — APPROVED** [review]

**Review by b1bart — APPROVED** [review]

**Pull request merged**